File: version4j/src/main/java/com/only4/algorithm/version4j/leetcode/TopKFrequent.java
package com.only4.algorithm.version4j.leetcode;

import java.util.HashMap;
import java.util.Map;
import java.util.PriorityQueue;

public class TopKFrequent {
    public int[] topKFrequent(int[] nums, int k) {
        // 1. 统计频率
        Map<Integer, Integer> freqMap = new HashMap<>();
        for (int num : nums) {
            freqMap.merge(num, 1, Integer::sum);
        }

        // 2. 使用最小堆（按频率升序），维护 size 为 k
        PriorityQueue<Map.Entry<Integer, Integer>> minHeap =
                new PriorityQueue<>(Map.Entry.comparingByValue());

        for (Map.Entry<Integer, Integer> entry : freqMap.entrySet()) {
            minHeap.offer(entry);
            if (minHeap.size() > k) {
                minHeap.poll(); // 删除频率最小的
            }
        }

        // 3. 提取结果（注意堆中元素不保证顺序）
        int[] result = new int[k];
        for (int i = k - 1; i >= 0; i--) {
            result[i] = minHeap.poll().getKey();  // 倒着取出来可选，保证从大到小
        }

        return result;
    }
}

File: version4k/src/main/kotlin/com/only4/algorithm/version4k/leetcode/topKFrequent.kt
package com.only4.algorithm.version4k.leetcode

import java.util.*

fun topKFrequent(nums: IntArray, k: Int): IntArray {
    // 统计每个元素出现的频率
    val frequencyMap = mutableMapOf<Int, Int>()
    nums.forEach { num ->
        frequencyMap[num] = frequencyMap.getOrDefault(num, 0) + 1
    }

    // 如果k等于唯一元素的数量，直接返回所有元素
    if (k == frequencyMap.size) {
        return frequencyMap.keys.toIntArray()
    }

    // 使用最小堆，按照频率排序（频率小的在堆顶）
    val minHeap = PriorityQueue<Pair<Int, Int>>(compareBy { it.second })

    // 维护大小为k的最小堆
    for ((num, frequency) in frequencyMap) {
        minHeap.offer(Pair(num, frequency))
        if (minHeap.size > k) {
            minHeap.poll() // 移除频率最小的元素
        }
    }

    // 将堆中的元素转换为数组
    return IntArray(k) {
        minHeap.poll().first
    }
}

